landoapi/models/revisions.py

```diff
@@ -32,6 +33,39 @@ class DiffWarningGroup(enum.Enum):
     LINT = "LINT"
 
 
+class Revision(Base):
+    """A Lando revision mapping to a Phabricator revision."""
+
+    # revision_id and diff_id map to Phabricator IDs (integers).
+    revision_id = db.Column(db.Integer, nullable=False, unique=True)
+    diff_id = db.Column(db.Integer, nullable=False)
+
+    # The actual patch.
+    patch_bytes = db.Column(db.LargeBinary, nullable=True)
+
+    # Patch metadata, such as author, timestamp, etc...
+    patch_data = db.Column(JSONB, nullable=False, default=dict)
+
+    def __repr__(self):
+        """Return a human-readable representation of the instance."""
+        return (
+            f"<{self.__class__.__name__}: {self.id} "
+            f"[D{self.revision_id}-{self.diff_id}]>"
+        )
+
+    @classmethod
+    def get_from_revision_id(cls, revision_id: int) -> "Revision":
+        """Return a Revision object from a given ID."""
+        return cls.query.filter(Revision.revision_id == revision_id).one_or_none()
+
+    def set_patch(self, raw_diff: bytes, patch_data: dict[str, str]):
+        """Given a raw_diff and patch data, build the patch and store it."""
+        self.patch_data = patch_data
+        patch = build_patch_for_revision(raw_diff, **self.patch_data)
+        self.patch_bytes = patch.encode("utf-8")
+        db.session.commit()
```

Comment:
I think if we were setting those fields directly in the caller, that would make sense, however since we doing more than just setting a field here, I don't see why this is not a good approach. I think it is better to keep database commits near where the model fields are being set. Otherwise, we could just move all of this functionality to where the caller is.

Reply:
Sorry, I wasn't clear enough with my statement. I wasn't trying to say this is not a good approach, I'm saying we shouldn't be mixing approaches if we can help it.

The concern I have is that sometimes we're committing the session in the API handler and sometimes we're committing the session in a function on the model. It would be better if we were either committing at the API layer consistently or in the model layer consistently. So we could commit as part of the caller (as I've suggested here, mostly because it would be the quickest change), or we could commit in the `Revision` model layer more consistently. For example instead of setting the diff ID and committing in the API handler we could create a `set_diff` function that both sets the diff and commits the session in the model.

Reply:
Probably what would be nice here is something similar to what Django's ORM does, e.g. a model-specific `save` method, but I think this is out of scope for this PR. I've moved the DB commit to the other module. I agree that the way DB sessions are handled in Lando are a little ad-hoc at this time.